### DISCUSSION TITLE
How to verify if setup support GPU acceleration?

### DISCUSSION THREAD

**Discussion by DynV** [ImageMagick/ImageMagick#2516] [Help]
I plan to run something thousands of time, and my system GPU is both more recent and in a higher price bracket (at each purchase time), so I'd like to put as much processing on it if possible. My ImageMagick instance is on [Cygwin](http://cygwin.com/) 3.1.6-1 and is version 7.0.10-27 Q16 x86_64 2020-08-14; it may not be the most recent, Cygwin's latest isn't always the software latest (read: many times).

How can one to verify ImageMagick setup support GPU acceleration?

Thank you kindly

**Comment by snibgo** [comment]
> How can one to verify ImageMagick setup support GPU acceleration?

This is really two questions.

Q1. Does this build of IM support OpenCL?

Answer: "magick -version" should show "Features: ... OpenCL". If OpenCL isn't listed, then the answer is "no".

Q2. Does IM's OpenCL use the GPU on this computer?

If the GPU isn't available, IM will automatically fallback to use the CPU. As far as I know, there are no diagnostics that say this has happened.

**Reply by DynV** [reply]
With
> $ magick -version
> Version: ImageMagick 7.0.10-27 Q16 x86_64 2020-08-14 https://imagemagick.org
> Copyright: © 1999-2020 ImageMagick Studio LLC
> License: https://imagemagick.org/script/license.php
> Features: Cipher DPC HDRI OpenMP(4.5)
> Delegates (built-in): bzlib cairo fftw fontconfig fpx freetype gslib heic jbig jng jp2 jpeg lcms lzma pangocairo png ps rsvg tiff webp x xml zlib

I guess there's no need to go into #2, but thanks again for answering. I--think--if I start running it when I go to sleep, it will be done when I wake up; if it's not enough, it becomes recurring and there's no not-too-complicated way to do it on Windows, I think I would try to find out if a live-nix could handle it, and if not if I want to make the system multi-OS.

**Comment by snibgo** [comment]
I recently downloaded and installed the pre-built binary ImageMagick-7.0.10-27-Q16-HDRI-x64-dll.exe . It shows:

```
Version: ImageMagick 7.0.10-27 Q16 x64 2020-08-10 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 192628806
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib
```
So I have the same version number as you, but a different set of features.

The feature list is defined at the IM build, not at runtime. I don't know why I have OpenCL but you don't. Did you install a pre-build binary, or build it yourself?

**Reply by DynV** [reply]
It's a pre-built binary. Cygwin installer is a GUI similar to what would be aptitude (if it was made GUI with everything about the same place).

**Comment by snibgo** [comment]
> Cygwin installer is a GUI ...

Does that mean you installed from the Cygwin installer? That may be the problem. I expect that version was built by the Cygwin team, not the ImageMagick team.

I suggest you go to https://www.imagemagick.org/script/download.php#windows and download one of the .exe files  and run it, and follow the dialogue to install ImageMagick.

**Comment by DynV** [comment]
As can be seen in
> $ /cygdrive/C/Program\ Files/ImageMagick-7.0.10-Q16-HDRI/magick.exe --version
> Version: ImageMagick 7.0.10-29 Q16 x64 2020-09-05 http://www.imagemagick.org
> Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
> License: http://www.imagemagick.org/script/license.php
> Visual C++: 192729111
> Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
> Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib

It was installed and I ran the same command as the Cygwin one with the same result. To go back to your 1st post in this thread. Is there no way to test point #2? Perhaps by me doing something only the CPU can do in ImageMagick then comparing it to something that should be done in GPU (if the installation was done correctly) and compare the execution times?

**Reply by fmw42** [reply]
You have to compare "apples to apples". See https://imagemagick.org/www/script/opencl.php. Run a large -gaussian-blur as a test and time the results when OpenCL is installed.  Then re-compile ImageMagick without OpenCL and time again.  Not sure if you can even disable OpenCL, though.

**Reply by DynV** [reply]
Using the executable in my previous post, it ended up with [FO4 - Loriya - new_perk_comparison-200906.txt - creation command beginning in Task Manager - Performance](https://imgur.com/a/LE2JQPL). Some GPU did end up being used for 2D... when opening some compared files in Windows default image viewer.

About
> I--think--if I start running it when I go to sleep, it will be done when I wake up

in my 2nd post (in this thread), it ended up being so (done when I woke up).
